### PR TITLE
The attack of the binding operators: introduce (let@) in the compiler codebase

### PR DESCRIPTION
**Pull request by gasche**
This PR introduces a `(let@)` binding operator that is the most trivial, but also arguably one of the most useful, binding operator: `let@ x = a in b` is just `a @@ fun x -> b`. (The name comes precisely from the `@@` operator.)

The operator is introduced in the compiler codebase, not in the standard library. (See, it's cute and harmless.)

I have rewritten most use-sites of `a @@ fun x -> b` to use it instead. Rationale:
- this gives an idea of the style we get when using it
- people who are allergic to fancy notations already hated the previous style, so maybe they won't complain less about the change (one can always dream)

The result looks somewhat similar to the [`use` keyword](https://docs.microsoft.com/en-us/dotnet/fsharp/language-reference/resource-management-the-use-keyword) in F# to explicitly allocate lexically-scoped resources. One can keep in mind that `let@` has a dynamic effect that scopes over the body, or abstract over it to see simple-looking code that reads nicely and does the right thing.

Examples (from the PR):

```ocaml
let interface ~source_file ~output_prefix =
  let@ info = with_info ~source_file ~output_prefix ~dump_ext:"cmi" in
  Compile_common.interface info


  let@ () = Profile.record_call "parsing" in
  file_aux ~tool_name inputfile parse invariant_fun kind


let copy_file src dest =
  let@ ic = with_input_file ~bin:true src in
  let@ oc = with_output_file ~bin:true dest in
  copy_chan ic oc
```

I found this operator very useful when working with simplified forms of continuation monads, for example as part of my research work on the [Inferno](https://gitlab.inria.fr/fpottier/inferno) library. It may take some time getting used to (as for any new notation), but I believe that it will replace the `@@ fun x ->` notation as people get more used to binding operators.

Note: some people have proposed in the distant past to use exactly this definition for a "flexible monadic-bind", usually with `let!` as the notation.

Note: There are some callsites in ocamltest that use the `@@ fun` notation with non-unary currified functions: `Action.make [...] @@ fun log env ->`. The callsites have not been converted: they would require to tuplify the function, which is used (without `@@`-notation) in several other places of the codebase. In the future when designing new quantifier-like operations, it is recommended to make their continuation tuplified instead of currified to ease binding-operator usage.

### CODE REVIEW

**Comment by gasche** [comment]
I briefly wondered about adding a corresponding `and@` operator (for use in the `copy_file` example above), but the definition is such that `let@ x1 = a1 and@ x2 = a2 in b` is defined as `let@ x1 = a1 in let@ x2 = a2 in b`, so there is no obvious benefit. There would be an naturalness argument if non-recursive `let .. and` was common in OCaml codebases, but it's generally not (at least not in the compiler). So I went without this `(and@)` operator for now.

**Comment by yawaramin** [comment]
I like it but, if I may be allowed to bikeshed a bit, I like `let&` for this, because it is reminiscent of 'borrowing' a la Rust: https://discuss.ocaml.org/t/downsides-to-calling-gc-full-major-at-exit/5570/16

**Comment by hcarty** [comment]
I like this a lot overall! This is tempting to use in work and personal projects, particularly when multiple resources using nested `with_x` functions are involved. 

Calls to `Fun.protect` with this syntax look a little odd. They look like side effect only function calls like `let () = print_endline "Hello" in e`. It's probably still worth using the `let@` in those cases though for consistency, but when reading the diff those were the lines that caused me to go back and read them again.

**Comment by lpw25** [comment]
I agree that a binding operator is no worse than using `@@` with `with_foo` functions. However, it suffers from the same obvious drawback: it takes a construct whose entire purpose is to delineate a region of code and makes the end of that region implicit and unclear. There will be bugs caused by using this style. This is in addition to the bugs like #9385 caused by the inherent flakiness in the `with_foo` pattern itself in the presence of currying.

Since this PR is no worse than the existing code in this regard, I have no issue with it being merged, but I would urge people to consider writing their code more robustly.

**Comment by bluddy** [comment]
I would prefer to see this introduced to the ecosystem and experimented with before the compiler adopts it. I don't find the examples in the changed files more readable. Instead, the reader now has to remember what is a currently unique usage of a new feature -- one that was introduced for monadic code, no less.

**Comment by gasche** [comment]
> I would prefer to see this introduced to the ecosystem and experimented with

I defined this operator and used it in the Inferno codebase. Does this count?

> before the compiler adopts it

Why?

Remember that this proposal only touches the compiler codebase, it does *not* add anything in the standard library and does not change anything for users of the compilers, just for people working on the compiler codebase.

**Comment by bluddy** [comment]
It just seems premature to me to introduce constructs into the compiler codebase when they haven't become popular in the ecosystem. It makes the codebase harder to understand, because we don't have the built-up knowledge from dealing with it elsewhere. If this construct never becomes popular in the ecosystem, it'll be a unique one-off (except for the other codebase you mentioned). I can only speak for my own experience, but since I know these constructs were added for monadic computation, I automatically seek that pattern. `@@` is something I'm used to, and have no problem parsing. Just my opinion though.

**Comment by xavierleroy** [comment]
(We have 250 open pull requests and the window is closing for 4.12, but let's argue about trivial rewrites in the compiler code base.)

I have no problems with either `a @@ fun x -> b` or `let@` to write code in continuation-passing style.  However, like @lpw25 I would not use either notation for higher-order functions such as `with_file_output`, because their function argument is not a continuation and it matters very much where the function ends.  

In other words: the binding operators are great for monads and applicative structures, but we should think twice before using them in other contexts.

**Comment by gadmm** [comment]
@lpw25 Do you have examples in mind where the end of the scope induced by "let..in" is hard to determine? This is a genuine question, I remember I used to be surprised sometimes by the parsing in other contexts. In C++ and Rust, programmers have no problem determining where the implicit destructor is run and it is quite easy to get used to, whereas I agree that in OCaml ambiguities could be an issue if they exist.

I have no opinions about whether it should enter the compiler codebase but I agree it would be nice if the community experimented with it. A specific notation `let&` one reserves for the with_ idiom rather than a generic `let@` could also be the occasion to teach the pitfalls and useful for drawing the reader's attention that something happens at the end of scope.

**Comment by lpw25** [comment]
> Do you have examples in mind where the end of the scope induced by "let..in" is hard to determine?

I don't think there is anywhere where it is genuinely ambiguous, not like trying to work out where an `if` statement ends or anything. It is still much less clear than having an explicit delimiter though.

Note that the model of "value destroyed when it goes out of scope" only applies for some of the functions converted by this PR. It makes little sense for others like `Fun.protect`.

**Comment by gadmm** [comment]
So it would be dangerous to put everything under the same generic binder `let@` simply acting as a replacement of `@@`, but I believe in the viability of a separate `let&` idiom to make a distinction about the special case of resources, if used appropriately, to replace the explicit form. Explicit delimiters can also be cumbersome with the indentation, and (IIUC your comment) they do not seem to give more information to the accustomed reader. This PR advocates an analogous of the `use` keyword from F# so this might be an idea to explore.

**Comment by ivg** [comment]
@gasche, my main concern is whether `let@` prevents the closure creation the same as `@@` and `|>` do? 

> Do you have examples in mind where the end of the scope induced by "let..in" is hard to determine?

```ocaml
let my_monadic_function args f =
   with_file name @@  fun input -> 
   fresh_variable () @@ fun var ->
   f input var
```

The same using binding operators

```ocaml
let my_monadic_function args f =
   let@ input = with_file name in
   let@ var = fresh_variable () in
   f input var
```

The issue with this code is that `f` is run in the dynamic context of the `with_file` function and the channel escapes the `my_monadic_function` and got caught in the monadic computation chained with it. Therefore it is nearly impossible to find the place where it will actually be released. 

To summarize, RAII-style functions are not monadic computations and shall not use the monadic/binding style operators as this style composes computations on the same level (which is made syntactically clear with indentation). It is better to use a keyworded argument to make their scope more explicit,

```ocaml
let my_monadic_function args f =
   let input = with_file name ~f:(fun lines -> 
      ...
      result) in
   let@ var = fresh_variable () in
   f input var
```

File: debugger/primitives.ml [hunk 1]
```diff
@@ -26,6 +26,8 @@ let cleanup e f =
 
 let nothing _ = ()
 
+let ( let@ ) f x = f x
```
Comment (ivg) [inline debugger/primitives.ml:29]: I wonder if it should be
```ocaml
external ( let@ ) : ('a -> 'b) -> 'a -> 'b = "%apply"
```

**Comment by gadmm** [comment]
@ivg I recognise @lpw25's point in your example, both the `@@` and `let@` versions are equally bad for readability. However it is not an example where the lexical scope of the variable `input` is hard to determine. With an idiom-specific binding `let&`, programmers would learn that the following:
```ocaml
let my_monadic_function args f =
  let& input = with_file name in
  fresh_variable () @@ fun var ->
  f input var
```
shall ring the same alarm bells in their minds as the following:
```ocaml
let my_monadic_function args f =
  with_file name (fun input ->
    fresh_variable () @@ fun var ->
    f input var)
```
In all cases monadic expressions easily lead to letting values escape.

**Comment by gasche** [comment]
I disagree with the main thrust of the comments:

1. To me the argument of `with_file` *is* a continuation. (It is not used in tail position within the function itself, so `with_file` is not the CPS translation of a "pure" function... obviously because otherwise we would just write the direct version.)
2. I find the warnings about scope ambiguities unconvincing. We know what the scope of a `let .. in ..` is, these operators behave in the same way. I reviewed all changes I did in the PR over the compiler codebase, I don't find any case where it introduced some sort of scope ambiguity. On the contrary, I think that making the code easier to read (as in this PR) makes eventual scoping issues easier to spot, while making the code harder to read (by imposing parentheses etc.) make the code harder to understand, including to spot the issue.

One comment about `let@` vs. monadic bind: one notable difference is that for an arbitrary continuation-applying `let@`, the monadic associativity law may not hold:

```ocaml
let@ bar =
  let@ foo = use ... in
  use (... bar ...)
in
baz

(* may differ from *)
let@ foo = use ... in
let@ bar = use ( ... foo ... )  in
baz
```

At first I thought some of the arguments here amounted to saying that we naturally expect associativity to hold of `let`-constructs, so it would be bug-inducing to use a "let" notation that is not associative. But then again, looking at the two examples above: to me it is very clear that the scoping of `e2` is not the same in both examples, and I don't see how this notation makes it easier to mistakenly use the second form instead of the first form. If anything, I think that it is mildly easier to spot the issue, or in general to understand what is going on, than if you write

```ocaml
use ... ~f:(fun foo ->
  use ( ... foo ... ) ~f:(fun bar ->
    blah
  )
)

(* instead of *)

use
  (use ... ~f:(fun foo ->
    ... foo ...
  ))
  ~f:(fun bar ->
    blah
  )
```

just because the code is easier to read.

**Comment by gasche** [comment]
The value that *I* see in the PR is that it makes code easier to read. I think this will be especially important if we want to write some parts of the compiler in continuation-passing style, to avoid the dreaded stack overflows on large/unnatural codebases. I don't really care for the naming (I am happy with `let&` if people prefer it, as long it can be used freely and not just for some constrained notion of resource-declaration).

**Comment by ivg** [comment]
I agree that the PR doesn't introduce any new ambiguities and probably makes the code more readable. 
I said "probably", because in the end, it is subjective what is easier to understand, e.g.,
```
let@ () = with_something in
foo
```
vs.
```
with_something @@ fun () -> 
foo
```

My main concern with this PR is that we are switching from `@@` which is an `apply` primitive to `let ( let@ ) f x = f x`, which will introduce closures. Do you have any thoughts on that? Is there any noticeable impact?


Concerning the scoping of the `with_xxx` syntax, this is purely a style issue. This PR doesn't change the style, and the issue that I was pointing, applies equally to both syntaxes, `@@ fun () ->` and `let@ () =`.  So, since this style is already in the code base, I don't see any more reasons to discuss it in the scope of this PR. A new PR could be created to address this issue if someone thinks that `with_file` introduces ambiguities.

**Comment by gasche** [comment]
I'm fine with defining `( let@ )` as `"%apply"` if that works, but I'm also 100% sure that it makes absolutely no performance difference, because all use-cases of `let@ x = foo in ...` use a `foo` that is much slower / memoy-hungry than allocating a closure.

> Concerning the scoping of the with_xxx syntax, this is purely a style issue. This PR doesn't change the style, and the issue that I was pointing, applies equally to both syntaxes [...]

Actually there is one place in the patch where I removed explicit parentheses to replace it with a `let@`:

```ocaml
  Misc.try_finally
    ~always:(fun () -> close_out oc)
    ~exceptionally:(fun () -> Misc.remove_file cmofile)
    (fun () ->
       bytecode
       |> Profile.(record ~accumulate:true generate)
         (Emitcode.to_file oc i.module_name cmofile ~required_globals);
    )
```

became

```ocaml
  let@ () =
    Misc.try_finally
      ~always:(fun () -> close_out oc)
      ~exceptionally:(fun () -> Misc.remove_file cmofile)
  in
  bytecode
  |> Profile.(record ~accumulate:true generate)
    (Emitcode.to_file oc i.module_name cmofile ~required_globals)
```

Now that I look at it again, I think that the following would be more readable than either versions:

```ocaml
  let@ () =
    Misc.try_finally
      ~always:(fun () -> close_out oc)
      ~exceptionally:(fun () -> Misc.remove_file cmofile)
  in
  let@ () = Profile.(record ~accumulate:true generate) in
  Emitcode.to_file oc i.module_name cmofile ~required_globals bytecode
```

but I'm happy to revert to the original form if people have a preference for it -- if they think it makes scoping of the `exceptionally` logic easier to understand?

**Comment by xavierleroy** [comment]
> I said "probably", because in the end, it is subjective what is easier to understand, e.g.,
```
let@ () = with_something in
foo
```
> vs.
```
with_something @@ fun () -> 
foo
```

My point is that both forms are bad because the extent of the `with_something` is ambiguous.  

What's wrong with lambda-calculus syntax?
```
with_something (fun () ->
   ...
)
```
or, for extra emphasis,
```
with_something begin fun () ->
  ...
end
```
?

**Comment by xavierleroy** [comment]
> The value that I see in the PR is that it makes code easier to read. I think this will be especially important if we want to write some parts of the compiler in continuation-passing style, to avoid the dreaded stack overflows on large/unnatural codebases.

1. I'm fine with let operators used for CPS and other monads, but not fine with the proposed uses involving resource management.
2. To avoid the dreaded stack overflows, I put more hopes in TRMC than in CPS conversion.  How's TRMC going, by the way?

**Comment by ivg** [comment]
> What's wrong with lambda-calculus syntax?

Nothing wrong. In fact, this is the style that I am using for resource management `with_` functions. I even prefer to use `begin/end` instead of `()` as it makes the scope of the resource more apparent.

**Comment by yallop** [comment]
I've made a small proposal for an alternative syntax `with e as x do ... done` in ocaml/RFCs#25

**Comment by gasche** [comment]
I had an interesting discussion with @gadmm today about this feature. He proposed an expansion of @ivg's example that I agree is a convincing argument against the current proposal:

```ocaml
open Lwt.BindingsOp (* let's assume this exists and binds Lwt.bind to `let*` *)
let my_monadic_function args f =
   let@ input = with_file name in
   let* data = Lwt.foo ... args in
   f input data

(* or, equivalently *)
let my_monadic_function args f =
  let@ input = with_file name in
  let@ data = Lwt.bind (Lwt.foo ... args) in
  f input data
```

The problem with this example (which was not so clear in @ivg's original example) is that `Lwt.bind`, or similar continuation operators, may "return" or "leak" their continuation (`fun data -> f input data`) to a scheduler instead of applying it directly, and return immediately before the continuation is executed (asynchronously). This is very incorrect here, as the `input` captured inside the continuation will be invalid after this returns happen, so the leaked continuation is already invalidated.

When we use `with_file`, we want to be sure that the function over the file handle will not leak the file handle; composing it with arbitrary control operators is dangerous and we should ensure that it does not belong to the path of least effort.

Note that this criticism applies no matter which concrete syntax is used:

```ocaml
let my_monadic_function args f =
  let@ input = with_file name in
  let@ data = Lwt.bind (Lwt.foo ... args) in
  f input data

let my_monadic_function args f =
  with_file name ~scope:(fun input ->
    Lwt.foo ... args >>= fun data ->
    f input data
  )

let my_monadic_function args f =
  with with_file name as input do
    with Lwt.bind (Lwt.foo ... args) as data do
      f input data
    done
  done
```

Those versions are all bad. This is not related to the let-versus-dodone choice, or let-versus-explicit-parentheses. `Lwt.bind` and `with_input` act as continuation operators with very different properties (`bind` leaks its continuation, `with_file` expects the guarantee that its continuation does not leak its input), and they cannot be composed safely.

**Comment by gasche** [comment]
This being said, I see three options that would improve over the current state, possible continuations (sic) for this PR.

## Two operators: arbitrary `let@`, restricted `let&`

Have a "unprincipled" binder for arbitrary continuation operators, `let@`, and a resource-specific binder `let&`. We should be careful to avoid nesting `let@` and `let&` together, or at least this should be a red flag to carefully review the effect of the `let@`-bound continuation operator.

In particular, we should avoid using `with_foo @@ fun v ->` and convert those to `let&` instead, to make this restriction clear.

(@gadmm suggests to enforce the usage of `let&` for resource operators by following the convention of labeling their continuation with a `scope:` label (`with_file path ~scope:(...)`). This is orthogonal to the general alternatives proposed here, but it suggests that there are API choices to nudge users in the right direction.)

The problematic example becomes:

```ocaml
let my_monadic_function args f =
   let& input = with_file name in
   let* data = Lwt.foo ... args in  (* or `let@ data = Lwt.bind (...) *)
   f input data
```
and here the mixing of `let&` with `let*/let@` is a code smell to investigate carefully. (Well, for `let*` this depends on whether the monad has a "leaky" bind.)

## Intermediate proposal: one principled `let@` operator

We could restrict `let@` by documenting that it must only be used for continuation operators that do not "leak" their continuation (it is only applied, possibly zero or several times). This is the largest class of operators that compose well with resource operators. Then using `let@ input = with_file foo` is safe/correct again, but `let@ data = Lwt.bind ...` is incorrect: if a monad has a "leaky" bind then it may not use `let@`.

This is my preferred approach. In particular:
- We can use `let@` in functions that mix continuation-passing-style and direct-style code for tailrec reasons. (This particular class of continuation operator is non-leaking, but it is also clearly not a candidate for `let&` usage.)
- With this proposal the code proposed in this PR is in fact completely fine, but more documentation on the "laws" (assumptions) of `let@` is required. (This is a conceptual change but it does not require a code change, as all envisioned uses of `let@` fit within this reasonable subclass.)

## Minimalist proposal: only `let&` 

A minimalist approach would be to offer `let&`, and no binding form for more general control operators. For the compiler today this proposal would be enough, given that almost all continuation operators, at least essentially-all touched in this PR, are resource operators. If later we want a binding operator for a different class of continuation operators, in particular the CPS usage I have in mind, we can discuss it separately and at that point chose something else -- under the constraint that it must compose well with `let&`.

**Comment by c-cube** [comment]
I came back to this issue after using `let@` in my own code a bit, and I must say, I absolutely love it for `with_…` functions.

A few things about people worried about scopes:

- the scope is not hard to predict, it's the same as any other `let`. If you want to make it clear you enter a new scope, do:
   ```ocaml
   begin
      let@ x = with_foo in
      let@ y = with_bar in
      ...
   end;

- languages with RAII have the exact same idiom, and RAII guards _do not_ introduce a new scope, nor do they add indentation. I think it's a good thing. Adding a new scope tends to push code to the right, and one might have a lot of RAII scopes in one function (instrumentation, resource handling, etc.)

- I think readability is a function of habit. Personally I find `let@` very readable, much better than `with_foo @@ fun x -> …` because we're used to seeing `let` already, and because it keeps the bound variable on the left.

That said, I don't have anything to say about `let@` _as used in the compiler_, I'd like to have it in `Fun` instead.

**Comment by c-cube** [comment]
I hope this can move forward. I've been using it even more, and it's just very very useful to me. I have it in scope everywhere now, and do `let@ () = Fun.protect ~finally:(…) in` in many places.

**Comment by yawaramin** [comment]
@c-cube unless I'm mistaken, this PR would not be useful to you directly, the operator is being introduced only internally. There's another PR for introducing let-operators into the standard library that has also been in limbo for a long time, that one would actually be user-visible.

**Comment by c-cube** [comment]
Ah, duh :facepalm: . Thank you @yawaramin . My testimonial remains :)

**Comment by gasche** [comment]
I think the feedback is useful, thanks!
- if people don't agree to have let@ internally, there is no chance to build consensus for stdlib inclusion, so "internal addition" is a good first step anyway
- the other PR does not consider this operator, only some applicative/monadic operators

**Comment by gasche** [comment]
Some more comments on this proposal:

- I still believe that we should have something like this in the compiler codebase to make higher-order wrappers more readable. We are starting to use those more and more, see for example #11536.
- One criticism of higher-order wrappers is that they are prone to mistakes related to partial application. (This is not specific to the letop syntax, it is an orthogonal problem.) I think that we should work on solving this problem, and we have a sketch of a proposal in #10604: an attribute to say that a polymorphic variable `'a` should be instantiated with a data type rather than a function type. (But I'm not sure if we can turn this intuition into something pragmatic that does not feel like a kludge. There is a risk we would try to do this "properly" as a sub-kind of something something and get lost in abstract type-theory land forever.)
- In the work of @lpw25 and @stedolan on "local" types, there is a way to type higher-order wrappers to prevent them from leaking their continuation: `local_ (... -> 'a) -> 'a` -- this point was made by @stedolan during a talk yesterday. This is related to the distinction I was making between `let&` and `let@` -- the former should be typed with a local continuation, the latter has no such restriction. I would still focus my proposal on the unprincipled `let@` as the first step in this direction for now.

**Comment by goldfirere** [comment]
I sadly do not remember why I was assigned this ticket, roughly 1 year ago. Was I to raise this as an issue for the language committee? Does anyone else remember?
